### DISCUSSION TITLE
Support passing of uvicorn specific config options to underlying runner

### DISCUSSION THREAD

**Discussion by anil-enable** [fastapi/fastapi-cli#13] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched in the documentation/README.
- [X] I already searched in Google "How to do X" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs/README and didn't find an answer.

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```Text
# errors with "No such option: --no-server-header"
fastapi dev src/app/main.py --port=8001 --no-server-header
```

### Description

uvicorn cli has a lot of config options allowing some degree of customization of the server, e.g. --no-server-header (switches off server header)

```bash
uvicorn app.main:app --port=8001 --no-server-header --reload
```
fastapi cli currently doesn't support these options. This blocks moving from uvicorn to fastapi cli where such customization of server is required

```bash
# errors with "No such option: --no-server-header"
fastapi dev src/app/main.py --port=8001 --no-server-header
```

If fastapi cli exposes a way to pass these options to uvicorn, it would be helpful.

### Operating System

macOS

### Operating System Details

macOS Sonoma Version 14.4.1

### Python Version

3.12.0

**Comment by EeJM** [comment]
We also need --forwarded-allow-ips=* for passing x-forwarded-for headers from a proxy in a Docker setup (from a non localhost IP).

**Reply by joke2k** [reply]
Seems to be a common request, great!

I'm using a function to create the app and it would be great to have this feature because i can use the --factory option.

**Comment by b4sen** [comment]
also the option to disable access logs would be useful

**Comment by sam-kleiner** [comment]
Also would be nice to pass `--reload-exclude` and `--reload-include` down to uvicorn

**Comment by lucafaggianelli** [comment]
We need `--no-server-header`, any update on this? happy to contribute with the PR

**Comment by kristk** [comment]
I would use a function like this to pass `--log-level` to uvicorn.

**Comment by jmehnle** [comment]
What about an option to invoke common ASGI servers other than uvicorn, such as hypercorn?

**Comment by sunil-enable** [comment]
@tiangolo Any thoughts on this thread. This will be quite useful to move completely to cli from current uvicorn command

**Comment by LoukaOctave** [comment]
Would love to see `--reload-dir` added to watch editable modules that aren't in the same path as my app

**Comment by nikstuckenbrock** [comment]
We would love to have the `log_config` option or another way to customize application logging.

**Comment by JoelYoung01** [comment]
I also need to pass an argument to --reload-dir and am not able using the fastapi cli, so I've ended up here. That alone would be nice, but a way to pass any arbitrary arg to the uvicorn run function would be nice.

**Comment by meschac38700** [comment]
I support this request, in fact, we lose a lot of uvicorn options which are very useful when using the fastapi cli to start a server.

> [!NOTE]
>The cli only uses the following options:
```Python
uvicorn.run(
    app=import_string,
    host=host,
    port=port,
    reload=reload,
    workers=workers,
    root_path=root_path,
    proxy_headers=proxy_headers,
    log_config=get_uvicorn_log_config(),
)
```
> [!WARNING]
> While we have all this
```Python
host: str = "127.0.0.1",
port: int = 8000,
uds: str | None = None,
fd: int | None = None,
loop: Literal["none", "auto", "asyncio", "uvloop"] = "auto",
http: Literal["auto", "h11", "httptools"] | Type[Protocol] = "auto",
ws: Literal["auto", "none", "websockets", "wsproto"] | Type[Protocol] = "auto",
ws_max_size: int = 16777216,
ws_max_queue: int = 32,
ws_ping_interval: float | None = 20.0,
ws_ping_timeout: float | None = 20.0,
ws_per_message_deflate: bool = True,
lifespan: Literal["auto", "on", "off"] = "auto",
interface: Literal["auto", "asgi3", "asgi2", "wsgi"] = "auto",
reload: bool = False,
reload_dirs: list[str] | str | None = None,
reload_includes: list[str] | str | None = None,
reload_excludes: list[str] | str | None = None,
reload_delay: float = 0.25,
workers: int | None = None,
env_file: str | PathLike[str] | None = None,
log_config: dict[str, Any] | str | RawConfigParser | IO | None = LOGGING_CONFIG,
log_level: str | int | None = None,
access_log: bool = True,
proxy_headers: bool = True,
server_header: bool = True,
date_header: bool = True,
forwarded_allow_ips: list[str] | str | None = None,
root_path: str = "",
limit_concurrency: int | None = None,
backlog: int = 2048,
limit_max_requests: int | None = None,
timeout_keep_alive: int = 5,
timeout_graceful_shutdown: int | None = None,
ssl_keyfile: str | PathLike[str] | None = None,
ssl_certfile: str | PathLike[str] | None = None,
ssl_keyfile_password: str | None = None,
ssl_version: int = SSL_PROTOCOL_VERSION,
ssl_cert_reqs: int = ssl. CERT_NONE,
ssl_ca_certs: str | None = None,
ssl_ciphers: str = "TLSv1",
headers: list[tuple[str, str]] | None = None,
use_colors: bool | None = None,
app_dir: str | None = None,
factory: bool = False,
h11_max_incomplete_event_size: int | None = None
```

**Comment by mhkarimi1383** [comment]
forwarded_allow_ips Is the most important configuration I think when running behind webserver in production, (Some of them can be set using envvar)

**Comment by aiguofer** [comment]
Came here because I want to use `--limit-concurrency`... The fact that there's no answer here makes me wonder... what's even the benefit of using the `fastapi-cli`?

**Comment by ingvaldlorentzen** [comment]
Would love this! We need `timeout_graceful_shutdown`

**Comment by NepNepFFXIV** [comment]
Would love this as well
